rbac/migration_tool/migrate.py

```diff
@@ -94,30 +96,28 @@ def migrate_role(
     return relationships, v2_role_bindings
 
 
-def migrate_users_for_groups(tenant: Tenant) -> list[common_pb2.Relationship]:
-    """Write users relationship to groups."""
-    relationships: list[common_pb2.Relationship] = []
-    for group in tenant.group_set.exclude(platform_default=True):
-        user_set: Iterable[Principal] = group.principals.all()
-        for user in user_set:
-            if (relationship := group.relationship_to_principal(user)) is not None:
-                relationships.append(relationship)
-    return relationships
+def migrate_groups_for_tenant(tenant: Tenant, replicator: RelationReplicator):
+    """Generate user relationships and system role assignments for groups in a tenant."""
+    groups = tenant.group_set.all()
+    for group in groups:
+        with transaction.atomic():
+            dual_write_handler = RelationApiDualWriteGroupHandler(
+                group, ReplicationEventType.MIGRATE_TENANT_GROUPS, replicator=replicator
+            )
+            if not group.platform_default:
+                dual_write_handler.generate_relations_to_add_principals(group.principals.all())
+            if group.system is False and group.admin_default is False:
+                system_roles = group.roles().filter(system=True)
+                if system_roles.exists() > 0:
+                    dual_write_handler.generate_relations_to_add_roles(system_roles)
+            dual_write_handler.replicate()
```

Comment:
We can address this in a follow-up PR if you would like, but along the lines of RHCLOUD-36014, I wonder if we should avoid calling this if we know there were no system_roles added and no principals added?

We could add that logic inside `replicate()`, but then it makes you wonder if we should really warn there? And in that case it's redundant because the outbox replicator already warns. So it feels like we should avoid calling replicate if we know there's nothing to replicate, otherwise the replicate method will be "suspicious" in some way :)

Reply:
yes, `.replicate()` should not be called when it is possible to determine that it is not needed otherwise we cannot capture unintended "empty" replication.  

I made changes here. Existence checks are moved outside the transaction block to avoid initializing the dual write handler when it is not needed. These checks are also not needed before the `generate_relations_XXX` method, as it they don't perform any work on empty arrays (system_role, principals)."